utils/lib/rr_graph/channel2.py

```diff
@@ -53,38 +52,74 @@ def __init__(self, tracks):
         self.trees = []
         self.tracks = sorted(tracks, key=lambda x: x[1] - x[0])
 
-    def _place_track(self, track):
-        """Add track to existing interval tree, if there is room (ie doesn't overlap)
-        If track won't fit in any existing tree, allocate a new tree for the track"""
-        for idx, tree in enumerate(self.trees):
-            if not tree.overlaps(track[0], track[1] + 1):
-                tree.add(
-                    Interval(begin=track[0], end=track[1] + 1, data=track[2])
-                )
-                return
+    def _start_track(self, track):
+        self.trees.append(list())
+        self.trees[-1].append(track)
 
-        self.trees.append(IntervalTree())
-        self.trees[-1].add(
-            Interval(begin=track[0], end=track[1] + 1, data=track[2])
-        )
+    def _add_track_to_last_tree(self, track, idx=-1):
+        self.trees[idx].append(track)
+
+    def _verify_trees(self):
+        for tree in self.trees:
+            for a, b in zip(tree, tree[1:]):
+                assert a[1] <= b[0]
 
     def pack_tracks(self):
         """pack all tracks"""
-        for track in self.tracks[::-1]:
-            self._place_track(track)
+
+        by_low = {}
+
+        def pop(low):
+            track = by_low[low].pop()
+
+            if len(by_low[low]) == 0:
+                del by_low[low]
+
+            return track
+
+        for low, high, key in self.tracks:
+            if low not in by_low:
+                by_low[low] = []
+
+            by_low[low].append((high, key))
+
+        if len(by_low) > 0:
+            high = max(by_low)
+
+        while len(by_low) > 0:
+            track_low = min(by_low)
+            track_high, key = pop(track_low)
+
+            self._start_track((track_low, track_high, key))
+
+            while True:
```

Comment:
```python
while track_high is not None:
```

otherwise line 96 would fail. Even if that couldn't happen, it's cleaner than `while True`.